x/ccv/provider/keeper/distribution.go

```diff
@@ -2,9 +2,80 @@ package keeper
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/x/ccv/provider/types"
 )
 
-func (k Keeper) GetFeeCollectorAddressStr(ctx sdk.Context) string {
+// EndBlockRD executes EndBlock logic for the Reward Distribution sub-protocol.
+// Reward Distribution follows a simple model: send tokens to the ConsumerRewardsPool,
+// from where they sent to the fee collector address
+func (k Keeper) EndBlockRD(ctx sdk.Context) {
+	// transfers all whitelisted consumer rewards to the fee collector address
+	k.TransferRewardsToFeeCollector(ctx)
+}
+
+func (k Keeper) RegisterConsumerRewardDenom(ctx sdk.Context, denom string, sender sdk.AccAddress) error {
+	// Send the consumer reward denom registration fee to the community pool
+	err := k.distributionKeeper.FundCommunityPool(ctx, sdk.NewCoins(k.GetConsumerRewardDenomRegistrationFee(ctx)), sender)
+	if err != nil {
+		return err
+	}
+	k.SetConsumerRewardDenom(ctx, denom)
+	k.Logger(ctx).Info("new consumer reward denom registered:", "denom", denom, "sender", sender.String())
+	return nil
+}
+
+func (k Keeper) GetConsumerRewardsPoolAddressStr(ctx sdk.Context) string {
 	return k.accountKeeper.GetModuleAccount(
-		ctx, k.feeCollectorName).GetAddress().String()
+		ctx, types.ConsumerRewardsPool).GetAddress().String()
+}
+
+func (k Keeper) SetConsumerRewardDenom(
+	ctx sdk.Context,
+	denom string,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.ConsumerRewardDenomsKey(denom), []byte{})
+}
+
+func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenoms []string) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.ConsumerRewardDenomsBytePrefix})
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()[1:]
+		consumerRewardDenoms = append(consumerRewardDenoms, string(key))
+	}
+
+	return consumerRewardDenoms
+}
+
+// TransferRewardsToFeeCollector transfers all consumer rewards to the fee collector address
+func (k Keeper) TransferRewardsToFeeCollector(ctx sdk.Context) {
+	// 1. Get the denom whitelist from the store
+	denoms := k.GetAllConsumerRewardDenoms(ctx)
+
+	// 2. Iterate over the whitelist
+	for _, denom := range denoms {
+		// 3. For each denom, retrieve the balance from the consumer rewards pool
+		balance := k.bankKeeper.GetBalance(
+			ctx,
+			k.accountKeeper.GetModuleAccount(ctx, types.ConsumerRewardsPool).GetAddress(),
+			denom,
+		)
+
+		// if the balance is not zero,
+		if !balance.IsZero() {
+			// 4. Transfer the balance to the fee collector address
+			err := k.bankKeeper.SendCoinsFromModuleToModule(
```

Comment:
What happens if denoms contains duplicated? Does the call to `SendCoinsFromModuleToModule` update the source balance so that the next call to `GetBalance` returns an empty balance?